### DISCUSSION TITLE
waiting for the request to be executed ? (flag)

### DISCUSSION THREAD

**Discussion by yuriy23** [eModbus/eModbus#314] [Q&A]
Good evening. there is a question... I make a plc, it scans LAD instructions. example on the screenshot. the master sends a command to the device, let's say 0x3 , the scan is done every 500ns. accordingly, the following requests are queued. is there a busy flag to skip these steps until the device responds or a timeout is triggered ?
![2023-09-22_19-20-32](https://github.com/eModbus/eModbus/assets/5504086/673a2f23-83e7-426c-b243-3ab2cea790a1)

**Comment by Miq1** [comment]
I am not sure what you are trying to explain with the screenshot, but if you really are trying to send a request every 500 **nano seconds** the queue will blow up in no time at all. I am not even aware of how this could be achieved with an ESP32.

If I was to make some sense from the remainder of your question, I would interpret that as "How can I detect if a request is still pending?" - did I get that right? If so, one easy way would be to limit the queue size to just 1. Then, whenever a request still is in processing, you will get an immediate "queue is full" (E8) error when trying to issue another request and can decide what to do in that case.

**Reply by yuriy23** [reply]
sorry for my English.. interpreter.. I'm sending a request to the slave. there are two options or a timeout if he did not respond or it takes time to respond, say 10ms. is there any flag to ignore the following requests? I think the queue option is rude. there must be a beautiful solution.

**Comment by Miq1** [comment]
Please try to explain your goal in more detail: would you like to send requests in a high frequency and normally it would take 10ms for the response to arrive, but when you will get a timeout, you would rather purge the queue from all requests that may have been issued in the meantime?

**Comment by yuriy23** [comment]
No , I 'll try to explain now

**Reply by yuriy23** [reply]
the scan of the program goes every 500ns , but the modbus is slow from the moment of sending to the response , let 's say 10 ms . I can make an external flag. if busy ==0 {err = MBslave1.addRequest(1, _addr, _funCode, _reg, _num); busy =1;} when responding, assign busy = 0; maybe there is a built-in mechanism?

**Reply by Miq1** [reply]
There is a method ``pendingRequests()`` that will give you the number of requests currently in the queue. You may call it as to use the number as a replacement for the ``busy`` flag you are considering.

But I am wondering again how you will send a request every 500ns... The CPU clock of the ESP32 normally is at 160MHz, thus a single clock tick takes about 6.3ns each. You would need to set up and send a request with a mere 80 ticks - impossible. Even 500us (micro seconds) will be tough, although being 1000 times slower!

**Reply by yuriy23** [reply]
S3 processor. of course not nc but microseconds. . this is the LAD scan rate.
that was the question, so as not to pull the modbus often, but only when the device responds.

**Reply by Miq1** [reply]
So the ``pendingRequests()`` method noted above will answer your question?

**Reply by Miq1** [reply]
As a note again: 500us between requests will not work at all with Modbus RTU, because the minimum possible quiet time between messages is 1700us. That goes twice (one pause after the request and the response both), so any poll rate below ~3500us will be forbidden by the Modbus standard.
Modbus TCP may work, but still does require a fast IP connection.

**Reply by yuriy23** [reply]
yes, I don't need to make requests every 500mcroseconds. on the contrary, I need to skip this instruction until the slave responds.

**Reply by Miq1** [reply]
If you could move the requests to a dedicated task, another alternative will be to use the ``syncRequest()`` calls there that will hold the task until a response is received or a timeout will happen.
